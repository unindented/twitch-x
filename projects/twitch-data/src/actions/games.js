import {getAction} from '../api'

export function loadTopGames ({limit}) {
  const action = {
    type: 'LOAD_TOP_GAMES'
  }
  return getAction(action, '/games/top', {query: {limit}})
}

export function loadSearchGames ({query, limit}) {
  const action = {
    type: 'LOAD_SEARCH_GAMES',
    payload: query
  }
  return getAction(action, '/search/games', {query: {query: encodeURIComponent(query), limit}})
}
