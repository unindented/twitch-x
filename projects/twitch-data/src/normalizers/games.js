import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const processGame = (game) => {
  return {
    id: valueOrNull(game.name),
    name: valueOrNull(game.name),
    image: valueOrNull(game.box.template)
  }
}

const topGame = new schema.Entity('games', {}, {
  idAttribute: (value) => value.game.name,
  processStrategy: ({game, viewers, channels}) => {
    return {
      ...processGame(game),
      viewers: valueOrNull(viewers),
      channels: valueOrNull(channels)
    }
  }
})

const searchGame = new schema.Entity('games', {}, {
  idAttribute: 'name',
  processStrategy: processGame
})

const rootTop = [topGame]
const rootSearch = [searchGame]

export function normalizeTopGames (payload) {
  return normalize(payload.top, rootTop)
}

export function normalizeSearchGames (payload) {
  return normalize(payload.games, rootSearch)
}
