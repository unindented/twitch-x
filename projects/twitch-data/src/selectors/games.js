import {createSelector} from 'reselect'

export const getTopGames = createSelector(
  (state) => state.data.games.top,
  (state) => state.data.games.byId,
  (topGameIds, gamesById) => topGameIds.map((id) => gamesById[id])
)

export const getSearchGames = createSelector(
  (state) => state.data.games.search,
  (state) => state.data.games.byId,
  (searchGameIds, gamesById) => searchGameIds.map((id) => gamesById[id])
)

export const getGame = createSelector(
  (state, props) => state.data.games.byId[props.id],
  (game) => game
)
