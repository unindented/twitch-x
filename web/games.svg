<?xml version="1.0" encoding="UTF-8"?>
<svg width="256px" height="256px" viewBox="0 0 256 256" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <defs></defs>
    <g id="Assets" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Games" fill-rule="nonzero" fill="#FFFFFF">
            <path d="M224,64 L32,64 C20.2666667,64 10.6666667,73.6 10.6666667,85.3333333 L10.6666667,170.666667 C10.6666667,182.4 20.2666667,192 32,192 L224,192 C235.733333,192 245.333333,182.4 245.333333,170.666667 L245.333333,85.3333333 C245.333333,73.6 235.733333,64 224,64 Z M117.333333,138.666667 L85.3333333,138.666667 L85.3333333,170.666667 L64,170.666667 L64,138.666667 L32,138.666667 L32,117.333333 L64,117.333333 L64,85.3333333 L85.3333333,85.3333333 L85.3333333,117.333333 L117.333333,117.333333 L117.333333,138.666667 Z M165.333333,160 C156.48,160 149.333333,152.853333 149.333333,144 C149.333333,135.146667 156.48,128 165.333333,128 C174.186667,128 181.333333,135.146667 181.333333,144 C181.333333,152.853333 174.186667,160 165.333333,160 Z M208,128 C199.146667,128 192,120.853333 192,112 C192,103.146667 199.146667,96 208,96 C216.853333,96 224,103.146667 224,112 C224,120.853333 216.853333,128 208,128 Z" id="Shape"></path>
        </g>
    </g>
</svg>