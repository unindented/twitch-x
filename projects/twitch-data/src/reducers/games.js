import update from 'immutability-helper'
import {normalizeTopGames, normalizeSearchGames} from '../normalizers/games'

const initialState = {
  byId: {},
  top: [],
  search: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_TOP_GAMES_SUCCESS': {
      const {entities, result} = normalizeTopGames(action.payload.response)
      return update(state, {
        byId: {$merge: entities.games},
        top: {$set: result}
      })
    }
    case 'LOAD_SEARCH_GAMES_SUCCESS': {
      const {entities, result} = normalizeSearchGames(action.payload.response)
      return update(state, {
        byId: {$merge: entities.games},
        search: {$set: result}
      })
    }
  }

  return state
}
