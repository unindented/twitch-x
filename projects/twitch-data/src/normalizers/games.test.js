import {normalizeTopGames, normalizeSearchGames} from './games'

const topGamesFixture = require('../__fixtures__/topGames.json')
const searchGamesFixture = require('../__fixtures__/searchGames.json')

describe('normalizers/games', () => {
  describe('.normalizeTopGames', () => {
    it('normalizes the payload', () => {
      const result = normalizeTopGames(topGamesFixture)
      expect(result).toMatchSnapshot()
    })
  })

  describe('.normalizeSearchGames', () => {
    it('normalizes the payload', () => {
      const result = normalizeSearchGames(searchGamesFixture)
      expect(result).toMatchSnapshot()
    })
  })
})
