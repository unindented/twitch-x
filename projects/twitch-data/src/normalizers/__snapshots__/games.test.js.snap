// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizers/games .normalizeSearchGames normalizes the payload 1`] = `
Object {
  "entities": Object {
    "games": Object {
      "SC2VN": Object {
        "id": "SC2VN",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/SC2VN-{width}x{height}.jpg",
        "name": "SC2VN",
      },
      "StarCraft": Object {
        "id": "StarCraft",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/StarCraft-{width}x{height}.jpg",
        "name": "StarCraft",
      },
      "StarCraft Ghost": Object {
        "id": "StarCraft Ghost",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/StarCraft%20Ghost-{width}x{height}.jpg",
        "name": "StarCraft Ghost",
      },
      "StarCraft II": Object {
        "id": "StarCraft II",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/StarCraft%20II-{width}x{height}.jpg",
        "name": "StarCraft II",
      },
      "StarCraft: Brood War": Object {
        "id": "StarCraft: Brood War",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/StarCraft:%20Brood%20War-{width}x{height}.jpg",
        "name": "StarCraft: Brood War",
      },
      "StarCraft: Retribution": Object {
        "id": "StarCraft: Retribution",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/StarCraft:%20Retribution-{width}x{height}.jpg",
        "name": "StarCraft: Retribution",
      },
    },
  },
  "result": Array [
    "StarCraft II",
    "StarCraft: Brood War",
    "StarCraft",
    "SC2VN",
    "StarCraft: Retribution",
    "StarCraft Ghost",
  ],
}
`;

exports[`normalizers/games .normalizeTopGames normalizes the payload 1`] = `
Object {
  "entities": Object {
    "games": Object {
      "Counter-Strike: Global Offensive": Object {
        "channels": 758,
        "id": "Counter-Strike: Global Offensive",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/Counter-Strike:%20Global%20Offensive-{width}x{height}.jpg",
        "name": "Counter-Strike: Global Offensive",
        "viewers": 393419,
      },
      "Hearthstone": Object {
        "channels": 280,
        "id": "Hearthstone",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/Hearthstone-{width}x{height}.jpg",
        "name": "Hearthstone",
        "viewers": 38773,
      },
      "League of Legends": Object {
        "channels": 1818,
        "id": "League of Legends",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/League%20of%20Legends-{width}x{height}.jpg",
        "name": "League of Legends",
        "viewers": 121668,
      },
      "PLAYERUNKNOWN'S BATTLEGROUNDS": Object {
        "channels": 1943,
        "id": "PLAYERUNKNOWN'S BATTLEGROUNDS",
        "image": "https://static-cdn.jtvnw.net/ttv-boxart/PLAYERUNKNOWN%27S%20BATTLEGROUNDS-{width}x{height}.jpg",
        "name": "PLAYERUNKNOWN'S BATTLEGROUNDS",
        "viewers": 76787,
      },
    },
  },
  "result": Array [
    "Counter-Strike: Global Offensive",
    "League of Legends",
    "PLAYERUNKNOWN'S BATTLEGROUNDS",
    "Hearthstone",
  ],
}
`;
